test/attr/attr_dynamic_member_lookup.swift

```diff
@@ -811,9 +811,16 @@ internal var rightStructInstance: SR12425_R = SR12425_R()
 
 public extension SR12425_R {
   subscript<T>(dynamicMember member: WritableKeyPath<SR12425_S, T>) -> T {
-      // TODO(Diagnostics): bad diagnostic for member assign.
-      // A better diagnostic would be: key path of type WritableKeyPath<SR12425_S, T> cannot be applied to a base of type SR12425_R
-      get { rightStructInstance[keyPath: member] } // expected-error {{cannot convert return expression of type 'Any?' to return type 'T'}}
-      set { rightStructInstance[keyPath: member] = newValue } // expected-error {{type of expression is ambiguous without more context}}
+      get { rightStructInstance[keyPath: member] } // expected-error {{key path of type 'WritableKeyPath<SR12425_S, T>' cannot be applied to a base of type 'SR12425_R'}}
```

Comment:
The problem with this diagnostic is that it doesn't really point out what _exactly_ went wrong here. I think we should add a note just like we do for generic arguments which would mention that `base is X and key path root is Y` or re-phrase the diagnostic.

Reply:
Humm, what about `key path of type 'WritableKeyPath<SR12425_S, T>' with root type 'SR12425_S' cannot be applied to a base instance of type 'SR12425_R'`? What do you think @xedin?

Reply:
the problem is not with `base` vs. `base instance` but instead I think that diagnostic should say that it couldn't use key path because its root is different from base it's applied on (in this example `SR12425_S` vs. expected `SR12425_R`). The type of the key path itself or its returned value doesn't really matter here IMO.

Reply:
Ahh true, so I think the note as you mention should be a good option

Reply:
I wonder whether diagnostic like `cannot apply key path with root type 'SR12425_S' to base of type 'SR12425_R'` would be a better diagnostic? WDYT? It seems to me that mentioning a particular type of key path might be contributing nothing in this situation...

Reply:
Sorry to bikeshed on this but I'm just trying to find a way to express information we are trying to covey without use of note and unrelated information...

Reply:
I agree, but I think we should create a new diagnostic because this is used for another case (AnyKeyPath is applicable) where it seems like show the key path type is important since the key path types are fully compared https://github.com/apple/swift/blob/a41a2ffb7ef21ca045722461465b0e2e2d86756c/lib/Sema/CSApply.cpp#L1720. 
But I can't think in a case where this here, where the failure is not based on the root type mismatch... do you think we should let this at it is or change the message in both cases?

Reply:
I think it's okay to add a new diagnostic for this. In the future we'd probably want to remove `cannot apply` diagnostic because it's too general to be useful.

Reply:
Just adjusted :)

Reply:
Thank you!